### ISSUE TITLE
Support MinReadySeconds on StatefulSet

### ISSUE THREAD

**Issue by grobie** [kubernetes/kubernetes#65098]
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind feature

**What happened**:

Using a StatefulSet for memcached clusters would be a good choice, as stable hostnames will avoid unnecessary data reassignment during a deploy. In order to let cache hit rate stabilize before updating the next pod, I was looking for the `MinReadySeconds` option as available in Deployment, DaemonSet, and ReplicaSet. Unfortunatley it doesn't look like that option exists in StatefulSets.

**What you expected to happen**:

Have a MinReadySeconds option in Statefulset.

**How to reproduce it (as minimally and precisely as possible)**:

https://github.com/kubernetes/kubernetes/blob/1a7a609cb1ee1c7746d1ebd01eb1262b106431a3/staging/src/k8s.io/api/apps/v1/types.go#L114-L173

**Anything else we need to know?**:

I'd be happy to contribute a patch, if you confirm the lack of MinReadySeconds on StatefulSets is not desired.

**Comment by grobie** [comment]
@kubernetes/sig-apps-feature-requests

**Comment by mattfarina** [comment]
/cc @enisoc @foxish @janetkuo @smarterclayton

Do we have any context why StatefulSet doesn't have this property but others do? Any thoughts on adding it?

**Comment by foxish** [comment]
IIRC, StatefulSet does not have the notion of "availability" in the same way that a deployment or replicaset has. It seems like a useful addition to me.

**Comment by enisoc** [comment]
For context, I think the reason StatefulSet doesn't have this is because MinReadySeconds at the workload controller level was considered deprecated at that time, but the replacement (MinReadySeconds at the Pod level) never materialized.

https://github.com/kubernetes/community/pull/478
https://github.com/kubernetes/community/pull/503#discussion_r109486980
https://github.com/kubernetes/kubernetes/issues/29229

**Comment by smarterclayton** [comment]
We should poke the pod level discussion.

**Comment by christianhuening** [comment]
@smarterclayton @foxish @enisoc @grobie i would also be interested in that

**Comment by grobie** [comment]
@smarterclayton @foxish @enisoc We just had another use case to control the prevent pods of a statefulset to be cycled too fast. Any updates here?

**Comment by krmayankk** [comment]
As per https://github.com/kubernetes/enhancements/pull/678#issuecomment-466291120 , it seems like we need both minReadySeconds and maxUnavailable  in StatefulSet to make it widely useful . I am actively working on adding the latter. Hopefully we can bring this (minReadySeconds) up in the next sig-apps meeting and take it further

**Comment by bgrant0607** [comment]
I also think minReadySeconds should be added to Pod, but I'm not opposed with adding it to StatefulSet first/also, since other controllers also have it

**Comment by Stono** [comment]
I would really like this feature.  At the moment we've had to change to `OnDelete` and manage the rollout outside of kubernetes

**Comment by sbernauer** [comment]
You can use the following workaround
```
 readinessProbe:
     exec:
         command:
             - "sh"
     initialDelaySeconds: 120
```

**Comment by grobie** [comment]
@sbernauer The need is to have a pod serving traffic for some time and warm up caches before continuing with the next pod. For that it needs to be ready already, so delaying the readiness probe doesn't help here.

**Comment by rdxmb** [comment]
Just an idea - maybe this is a workaround for you:

You can have two services. One quite "normal" and another with 
`publishNotReadyAddresses: true` . In my case, with minIO I have a headless service for that - where the different pods can talk to each other.
I have also set a `initialDelaySeconds: 60`

Another service (in my case with `ClusterIP`) is for routing the traffic from the clients.

When updating the statefulSet, each minIO-pod is waiting 60s before updating the next - even if they are ready before.

**Comment by ravisantoshgudimetla** [comment]
I think this has been a long due request from various kubernetes users. I agree that adding this field to pod might be beneficial in the long run, having said that as this never materialized I believe having that field in the statefulset is better  for now. We can have a follow-up to move this field to pod for every controller. I can work on this in 1.22 timeframe if @kubernetes/sig-apps-misc  folks are fine with the approach.

**Comment by smarterclayton** [comment]
Agree.  StatefulSets should be at least equivalent to DaemonSets and Deployments.  I would say we should add it to SS for now since it is a requirement to leverage minReadySeconds behind service load balancers to prevent killing pods in rotation before the new pods show up in rotation (i.e. normal SLB on clouds you have to wait long enough to see your new pod with maxUnavailable) .

**Comment by soltysh** [comment]
:+1: for my side as well, @ravisantoshgudimetla make sure to add that to one of the upcoming sig-apps meetings.

**Comment by ravisantoshgudimetla** [comment]
Done, added it to the agenda doc.

**Comment by andyxning** [comment]
We need to have a capability of expressing: "we need this pod to wait at least 30s before rolling update the next pod in StatefulSet".

**Comment by thockin** [comment]
Having just read the KEP (https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/2599-minreadyseconds-for-statefulsets#how-does-this-feature-react-if-the-api-server-andor-etcd-is-unavailable), I am not opposed to adding this for the purrpose of parity.  I'd really like to see it make its way down to pod so LBs can actually consume this.  But that is a different discussion.

**Comment by wilkejo** [comment]
Could someone elaborate why this is still open? For me it sounds like this is implemented in 1.25 as GA

**Comment by atiratree** [comment]
yes, this feature is already stable (https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#minimum-ready-seconds) and if there are any followups then they  should be discussed in a different issue or format

/close
